### ISSUE TITLE
Problem using LIME with XGBoost

### ISSUE THREAD

**Issue by beeva-josealberto** [thomasp85/lime#63]
Hi thomasp85, and thank you for this great package.

There is a problem when using XGBoost, specifically in the **predict_model.xgb.Booster** function. If the input to that function is a data.frame, the function executes:

`newdata <- xgboost::xgb.DMatrix(as.matrix(newdata))`

If **newdata** contains columns with type **factor**, then **as.matrix(newdata)** will be a character matrix, and **xgboost::xgb.DMatrix** will crash.

I may be wrong, but in my opinion there is a deeper problem here. In my particular case, I "one-hot encoded" the factor features before training the model. So what should the **explain** function input be? If I use the test set without one-hot encoding, **xgboost** will crash when trying to predict. On the other hand, if I use the one-hot encoded test set, will the permutations done by the package make sense?

In general, the LIME paper separates features used to train the model from human-understandable "variables" that will explain the model locally. How do we do that separation using this package?

------

**EDIT:** I am looking at the Python implementation. They let you enter the prediction function as a parameter. I think that would be a nice (flexible) solution to the problem. That way we can define the required transformations inside the custom prediction function.

-----

**EDIT2:** Ok, I did not read this:

> Out of the box lime supports models created using the caret and mlr frameworks. Support for other models are easy to achieve by adding a predict_model and model_type method for the given model.

This solves my problem. I'm sorry :)
